async-nats/src/jetstream/consumer.rs

```diff
@@ -15,19 +15,31 @@ use std::time::Duration;
 use serde::{Deserialize, Serialize};
 use time::serde::rfc3339;
 
-use crate::Error;
+use crate::{Error, Subscriber};
+
+use super::Context;
 
 pub trait IntoConsumerConfig {
     fn into_consumer_config(self) -> ConsumerConfig;
 }
 
 pub struct Consumer<T: IntoConsumerConfig> {
-    config: T,
+    pub(crate) context: Context,
+    pub(crate) config: T,
+    pub(crate) info: ConsumerInfo,
+    pub(crate) inbox: String,
+    pub(crate) subscription: Option<Subscriber>,
```

Comment:
Hmm, shared state? can't this live on `Batch`? 

```suggestion
    pub(crate) subscription: Option<Subscriber>,
```

Reply:
It should not. Creating new subscription for every batch is pretty bad for performance.

We can though switch to such approach as soon as muxing request subject is possible for Pull Consumers.
Would remove state and be more even more performant.

Reply:
Wouldn't be for every batch, this was partially why I proposed having iterators of iterators: to have somewhere that is not the consumer that can have proper mutability semantics while owning the state, performing the requests as needed etc.

Right now consumer acts like its an iterator semantically, (it's just not implementing the traits).
Quick abstract, the basic lowest level layer can semantically be something along the lines of the following:

```rust
// A fetch operation here is an iterator fetches batches (naming is hard).
pub struct Fetch {
  // owns the subscriber
  subscriber: Subscriber,
  // other relevant state.
}

impl Stream for Fetch {
    // Result to deal with errors coming from requests
    type Item = Result<Batch, Error>;
}

pub struct Batch {
  // take a reference etc, if you need it.
  subscriber: &'a Subscriber,
}

impl Stream for Batch {
  // Can't fail as far as I'm aware at this point, so no result.
  type Item = Message;

  fn poll_next(
         mut self: pin::Pin<&mut Self>,
         cx: &mut task::Context<'_>,
  ) -> task::Poll<Option<Self::Item>> {
    // Pull from the batch as needed.
  }
}

impl Consumer {
  pub fn fetch() -> Result<Fetch, Error> {
    Ok(Fetch {})
  }

  fn poll_next(
         mut self: pin::Pin<&mut Self>,
         cx: &mut task::Context<'_>,
  ) -> task::Poll<Option<Self::Item>> {
    // Dispatch requests as needed etc.
  }
}

fn main() {
  let consumer = ...;
  
  // Built-in combinators can flatten this again.  
  for batch in consumer.fetch().next().await? {
    for message in batch.next().await? {
      println!("{:?}", message);
    }
  }
```

Reply:
Just for clarity: we seem to find reasonable compromise with current implementation.